Add basic debian support

paluh opened the pull request:
I've added really basic `Apt` module. Unfortunately I have to use environment variable to force non interactive mode on `apt`, so I've added additional field for `env` to `Proc` type and `procEnv` constructor (I'm not really proud of this name).

I've tried to keep your coding style.

jimenezrick commented:
Hey, sorry for the delay, I'm looking at this. I like the possibility of passing env vars to the process, but I'll rename it to something like `procWithEnv`.

I'll fix a bit the style and make a couple of small changes but I'll merge it soon.

I was thinking at some point I need to start putting some kind of interfaces so you can have different implementations, such as for package management. You specify in the beginning of the script which implementations you want to use, i.e. apt, systemd, etc and then the rest of the script is generic.

paluh commented:
> I'll fix a bit the style and make a couple of small changes but I'll merge it soon.

Cool!

> I was thinking at some point I need to start putting some kind of interfaces so you can have different implementations, such as for package management. You specify in the beginning of the script which implementations you want to use, i.e. apt, systemd, etc and then the rest of the script is generic.

I know that in propellor they use type level sets to encode properties restricted to given system and to proof that "receipe" is consistent (regarding OS etc.) at compile time, but I'm not sure if it doesn't clutter API or complicates usage.
I was also thinking about building some extensible data type (a la Carte) of system properties above your core system, to allow translation into `Steps/Receipes` or allow interpretation which perform dry run (property checks) or allow property tree preprocessing (for example for some parallelism or efficiency - you can imagine aggregation of all `pacman -S/apt-get install` nodes etc.), but I'm not sure if I'm able to design anything serious... At first, in my spare time, I want to create some simple interpreter for `Steps` which would generate `Check` instead of `Receipe`.

Anyway, I really like clarity of your current design! I'm nearly sure that I'm going to migrate my current deployment from ansible "typeless configuration/variables soup" to your framework.

Thanks for this really cool project!

jimenezrick commented:
Have a look at my last commits, I added `withEnv` which I think it's a bit cleaner and I also added `Cook.Catalog.Debian.Apt` based on your code.

I had a very brief look at propellor in the past, I just prefer to have a very simple/lightweight framework that offer something better than Python/Bash, so not too obsessed with compile-time checks for this use case. Besides, I think using Haskell for this deployment/configuration stuff it's already a bit heavy weight, I personally prefer to install this `cook` lib in the target system and run the recipes in interpreted mode.

Having said that, not sure if my type `Recipe` is the best design, although is simple enough and works well for my use cases.

I like the idea of having system properties and checks, I guess that needs some design exploration. Let me know.

paluh commented:
> (...) I added withEnv which I think it's a bit cleaner and I also added Cook.Catalog.Debian.Apt based on your code.

Really nice. Thanks!

> I had a very brief look at propellor in the past, I just prefer to have a very simple/lightweight framework that offer something better than Python/Bash, so not too obsessed with compile-time checks for this use case.

Maybe this is a bit to heavy indeed. For me your solution have a good balance of simplicity and type safety, so maybe it's a good point to not adapt too fancy staff...

> Besides, I think using Haskell for this deployment/configuration stuff it's already a bit heavy weight, I personally prefer to install this cook lib in the target system and run the recipes in interpreted mode.

Hmmm, for me deployment with single executable seems really lightweight - compile, copy, run on multiple machines through ssh (with something like tmux-cssh). I really like that your focus is not on transport/execution layer.

> I like the idea of having system properties and checks, I guess that needs some design exploration. Let me know.

If I create something useful I will ping you back for sure!

I'm closing this pull request as it has nothing more to offer ;-)
